Custom CORS Error Header (#36)
SPSCommerce/sps-api-standards: standards/request-response.md

#### 403 Forbidden

> **Description**: The client is not authorized to access the resource, although it may have valid credentials. The API could use this status code in case business-level authorization failures. For example, the requestor does not have permission to request the organization's information.

Reviewer: ```suggestion
**Description**: The client is not authorized to access the resource, although it may have valid credentials. The API could use this status code to handle business-level authorization failures. For example, the requestor does not have permission to request the organization's information.
```

The usage of non-standard headers is not considered custom headers. For example, your API may consume or interact with other infrastructure outside of the contract specification for your service, such as `X-Forwarded-Host` or `X-Request-ID`, these are appropriate to continue using, but would not expect to identify them in your Open API spec for example. As such these types of headers not relevant to REST are intentionally left out of this documentation.
```

#### SPS-CORS-Error

**Type**: Response

**Support**: OPTIONAL

> **Description**: A [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) error with an invalid request on a non-200 status code can be difficult to understand the reasons why the request was rejected. Using a custom header provides the ability to include more detailed information to the consumer on why their CORS preflight OPTIONS request was rejected.

Reviewer: ```suggestion
**Description**: A [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) error from an invalid request will return a non-200 status code, but it can be difficult for an API consumer to understand the reason that their request was unsuccessful. A custom header provides more detailed information to the consumer on why their CORS preflight OPTIONS request was rejected.
```

- The header **MUST** be added when a `CORS` `OPTIONS` request is rejected.
- A response with this header **SHOULD** indicate a status code of `403 - Forbidden`.
- The header **MUST** contain one of the following values:

>     - __bad origin__ - Indicates to the consumer that the provided request header `Origin` header did not match approved list of source URLs.

Reviewer 2: ```suggestion
    - __bad origin__ - Indicates to the consumer that the provided request header `Origin` did not match approved list of source URLs.
```

Reviewer: Might be overly pedantic on my part, but I wonder if the words `domains` or `hostnames` might be a better choice than `URLs` here. URL makes me think that it can (and should) be constrained to a resource with a directory path.

Author: Love it, I'll hostnames to be consistent with terminology in URL Structure here https://spscommerce.github.io/sps-api-standards/standards/url-structure.html#overview

>     - __bad origin__ - Indicates to the consumer that the provided request header `Origin` did not match approved list of source URLs.

Author: ```suggestion
    - __bad origin__ - Indicates to the consumer that the provided request header `Origin` did not match the approved list of source hostnames.
```

- HTTP `OPTIONS` Method **MUST NOT** be used with any sensitive data.
- HTTP `OPTIONS` Method responses **MUST NOT** be intended to be cached.
- HTTP `OPTIONS` Method used for `CORS` integration **MUST** provide standard `CORS` headers for access control (`Access-Control-Allow-*`).

>     - Rejected `CORS` requests should return a `403` status code, including the [SPS-CORS-Error](#SPS-CORS-Error) custom header indicating the reason for rejection.

Reviewer 2: I can't tell if it's Github or if the indenting is off here

Reviewer 2: Oh wait, it might be indented intentionally :-D

Author: yep indentation intended in this case.
